### PR TITLE
Correct loss accumulation for grpo_fast

### PR DESCRIPTION
**Pull request by hamishivi**
Fixes up loss accumulation, see https://allenai.slack.com/archives/C08A1L22QMQ/p1764523483258119 for experiments.

Swaps default to tokenwise loss, which correctly divides the loss by the total number of valid tokens in the batch, rather then just in the current minibatch (incorrect when there are different numbers of tokens in each minibatch).

successful beaker run: https://beaker.org/ex/01KBBRN5HS1YYY6E2D4NFEGJC1

---

> [!NOTE]
> Switches GRPO loss to token-wise normalization across grad-accum groups and refactors masked_mean into rl_utils with new denominator handling.
> 
> - **GRPO training (open_instruct/grpo_fast.py)**:
>   - Introduces `Args.loss_denominator` ("token" or float) and validates via `utils.get_denominator`.
>   - Adds `calculate_token_counts` to precompute per-accumulation token totals for consistent normalization.
>   - Uses token-count denominator in `masked_mean` for policy/KL loss; removes per-minibatch scaling and multiplies by `world_size` when distributed.
>   - Refactors stats to new `masked_mean` signature and adds minor metric (`val/ratio_var`).
>   - Fixes minor typo in a comment.
> - **Utils (open_instruct/utils.py)**:
>   - Adds `get_denominator` helper with validation.
> - **RL utils (open_instruct/rl_utils.py)**:
>   - Moves `masked_mean` here (unchanged behavior) for shared use.
> - **Model utils (open_instruct/model_utils.py)**:
>   - Removes `masked_mean` (now in `rl_utils`).
> - **Tests**:
>   - Move `masked_mean` tests from `test_model_utils.py` to `test_rl_utils.py`.
>   - Add tests for `utils.get_denominator`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3c9720bca6c7f24f0d5e6c41c062c6945d00ee6b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/grpo_fast.py [hunk 1]
```diff
@@ -442,9 +444,17 @@ def __post_init__(self):
                 "use_vllm_logprobs sets old_logprobs to vLLM logprobs, making importance sampling pointless."
             )
         if self.masked_mean_denominator is not None:
-            assert self.masked_mean_denominator > 0, (
-                f"masked_mean_denominator (={self.masked_mean_denominator}) must be greater than 0!"
-            )
+            if isinstance(self.masked_mean_denominator, str):
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Can you change this block of code to be a function call? `get_denominator`, which combines all of this logic and has a few simple tests? (Using parameterized)

File: open_instruct/grpo_fast.py [hunk 2]
```diff
@@ -1117,6 +1124,33 @@ def train(
...
+                    for group_start in range(0, len(collated_query_responses), accumulation_steps):
+                        group_end = min(group_start + accumulation_steps, len(collated_query_responses))
+                        # Calculate local tokens for all minibatches in this accumulation group
+                        local_group_tokens = 0.0
+                        for i in range(group_start, group_end):
+                            mb_response_masks = collated_response_masks[i]
+                            mb_response_masks_bool = mb_response_masks[:, 1:].bool()
+                            if args.mask_tool_use and args.tool_use:
+                                mb_tool_mask = collated_tool_masks[i]
+                                mb_response_masks_bool = mb_response_masks[:, 1:].bool() & mb_tool_mask[:, 1:].bool()
```
[the remark this answers is not in the document]
Comment (finbarrtimbers) [reply open_instruct/grpo_fast.py]: I think the idea of a helper method is good tbh

File: open_instruct/grpo_fast.py [hunk 3]
```diff
@@ -252,8 +252,10 @@ class Args:
     """the length of the pack (you should prob set to the max length of the model)"""
     masked_mean_axis: int | None = None
     """the axis to compute the mean of the masked values"""
-    masked_mean_denominator: float | None = None
-    """Optional constant denominator for masked_mean; if set, divides by this instead of mask.sum"""
+    masked_mean_denominator: float | str | None = None
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Can you help me understand why we need this? Can we get rid of this and just do the right thing? 

What if we either: 

1. Get rid of this and always do "token"
2. have it default to "token" if no float is set?

File: open_instruct/grpo_fast.py [hunk 4]
```diff
@@ -1117,6 +1127,33 @@ def train(
...
+                        if dist.is_available() and dist.is_initialized():
+                            dist.barrier()
+                            local_group_tokens_tensor = torch.tensor(
+                                local_group_tokens, dtype=torch.float32, device=self.device
+                            )
+                            dist.all_reduce(local_group_tokens_tensor, op=dist.ReduceOp.SUM, group=None)
+                            accumulation_group_tokens[group_start] = local_group_tokens_tensor.item()
+                        else:
+                            accumulation_group_tokens[group_start] = local_group_tokens
+
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py:1200]: Can you makek this a function? 

```
accumulation_group_tokens = maybe_calculate_group_tokens(...)
```

File: open_instruct/grpo_fast.py (hunk 3 above)
Comment (hamishivi) [reply open_instruct/grpo_fast.py]: I'll set it to default to token! 
I think:
- having a float is useful since this is recommended by Dr GRPO
- The none is mainly to keep backwards compatible behaviour, but agreed could remove it. I'll run some comparisons and then remove.

File: open_instruct/grpo_fast.py [hunk 5]
```diff
@@ -498,6 +496,8 @@ def __post_init__(self):
                 self.gs_checkpoint_state_dir = f"{self.gs_bucket_path}/{beaker_users}/{checkpoint_dir_name}"
             else:
                 self.gs_checkpoint_state_dir = f"{self.gs_bucket_path}/{checkpoint_dir_name}"
+            if not checkpoint_dir_name.startswith("/filestore"):
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py:503]: Sorry to be nit picky, can we pull this into a separate PR? I can do it if you want.

File: open_instruct/grpo_fast.py [hunk 6]
```diff
@@ -1053,6 +1044,34 @@ def compute_logprobs(
 
         return collated_logprobs, collated_entropies
 
+    def calculate_token_counts(
+        self,
+        accumulation_steps: int,
+        collated_response_masks: list[torch.Tensor],
+        collated_tool_masks: list[torch.Tensor],
+    ) -> dict[int, float]:
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py:1056]: Add a docstring please!

File: open_instruct/grpo_fast.py [hunk 7]
```diff
@@ -1053,6 +1044,34 @@ def compute_logprobs(
...
+        accumulation_counts = {}
+        args = self.args
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: I don't think this (or device = self.device) are needed.

File: open_instruct/test_utils.py [hunk 8]
```diff
@@ -525,3 +525,14 @@ def mock_inspect_return(*args, **kwargs):
 #             "natolambert/tulu-v2-sft-mixture-science": 7468,  # original data slightly different
 #         }
 #         _ = get_datasets(dataset_mixer, splits=["train"], columns_to_keep=["messages"])
+
+
+@pytest.mark.parametrize("denominator,expected", [(None, None), ("token", "token"), (10.0, 10.0), (5, 5), ("5", 5)])
```
Comment (finbarrtimbers) [inline open_instruct/test_utils.py]: Can you make this a unittest.TestCase to be consistent?

File: open_instruct/utils.py [hunk 9]
```diff
@@ -2491,3 +2491,23 @@ def get_beaker_experiment_url() -> str | None:
...
+def get_denominator(masked_mean_denominator: float | str | None) -> float | str | None:
+    """Validate and return the masked mean denominator value."""
+    if masked_mean_denominator is None:
+        return None
+
+    if isinstance(masked_mean_denominator, str):
+        if masked_mean_denominator == "token":
+            return masked_mean_denominator
+        # Try to convert numeric strings to float
+        try:
```
Comment (finbarrtimbers) [inline open_instruct/utils.py]: I think it's fine to just directly do the conversion without the try/except.

File: open_instruct/grpo_fast.py [hunk 10]
```diff
@@ -1158,6 +1177,14 @@ def train(
             ratio_stats = torch.zeros(len(collated_query_responses))
             entropy_stats = torch.zeros(len(collated_query_responses))
             for epoch_idx in range(args.num_epochs):
+                # Pre-compute total tokens for each accumulation group if using "token" normalization
+                # This ensures all minibatches in an accumulation group are normalized by the same total
+                accumulation_token_counts = {}
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Can we do: 

if args.masked_mean_denominator == "token":
    accumulation_token_counts = self.calculate_token_counts(...)
else:
    accumulation_token_counts = {}

File: open_instruct/grpo_fast.py [hunk 11]
```diff
@@ -1167,6 +1194,14 @@ def train(
                     # if masking snippets, do it here.
                     if args.mask_tool_use and args.tool_use:
                         mb_response_masks_bool = mb_response_masks[:, 1:].bool() & mb_tool_mask[:, 1:].bool()
+
+                    # Determine the denominator for masked_mean normalization
+                    loss_denominator = args.masked_mean_denominator
+                    loss_axis = None
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Why do we need this variable if it's None? let's just inline it?

File: open_instruct/grpo_fast.py [hunk 12]
```diff
@@ -1272,6 +1307,12 @@ def train(
 
                     pg_loss_max = torch.max(pg_losses, pg_losses2)
 
+                    # for stats computation, for now no denominator is used
+                    # unless masked_mean_denominator is a numeric value.
+                    stats_denominator = (
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Can you move this to be defined right before it's used?

File: open_instruct/grpo_fast.py [hunk 13]
```diff
@@ -1163,6 +1187,19 @@ def train(
             ratio_stats = torch.zeros(len(collated_query_responses))
             entropy_stats = torch.zeros(len(collated_query_responses))
             for epoch_idx in range(args.num_epochs):
+                # Pre-compute total tokens for each accumulation group if using "token" normalization
+                # This ensures all minibatches in an accumulation group are normalized by the same total
+                accumulation_token_counts: dict[int, float] = {}
```
Comment (finbarrtimbers) [inline open_instruct/grpo_fast.py]: Remove this, as if we have an if/else?
